hibernate-core/src/main/java/org/hibernate/id/uuid/UuidVersion7Strategy.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.id.uuid;
+
+import java.security.SecureRandom;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.hibernate.engine.spi.SharedSessionContractImplementor;
+import org.hibernate.id.UUIDGenerationStrategy;
+
+import static java.time.Instant.EPOCH;
+import static java.time.temporal.ChronoUnit.MILLIS;
+
+/**
+ * Implements UUID Version 7 generation strategy as defined by the <a href="https://datatracker.ietf.org/doc/html/rfc9562#name-uuid-version-7">RFC 9562</a>.
+ *
+ * <ul>
+ *     <li>48 bits - 48-bit big-endian unsigned number of the Unix Epoch timestamp in milliseconds.</li>
+ *     <li>4 bits - version field, set to 0b0111 (7).</li>
+ *     <li>
+ *         12 bits - sub-milliseconds part of timestamp (resolution approximately 1/4 of millisecond)
+ *         to guarantee additional monotonicity.
+ *     </li>
+ *     <li>2 bits - variant field, set to 0b10.</li>
+ *     <li>14 bits - counter to guarantee additional monotonicity, resets to 0 when timestamp changes. </li>
+ *     <li>48 bits - pseudorandom data to provide uniqueness.</li>
+ * </ul>
+ *
```

Comment:
```suggestion
 *
 * @apiNote A Version 7 UUID features a time-ordered value field derived from the widely
 * implemented and well known Unix Epoch timestamp source, the number of milliseconds
 * since midnight 1 Jan 1970 UTC, leap seconds excluded.
 *
```